File: iphone/Classes/AccessibilityModule.m
/**
 * Titanium SDK
 * Copyright TiDev, Inc. 04/07/2022-Present. All Rights Reserved.
 * Licensed under the terms of the Apache Public License
 * Please see the LICENSE included with this distribution for details.
 */

#ifdef USE_TI_ACCESSIBILITY

#import "AccessibilityModule.h"
#import <TitaniumKit/TiUIView.h>
#import <TitaniumKit/TiUtils.h>

@implementation AccessibilityModule

#pragma mark - Internal

- (NSString *)apiName
{
  return @"Ti.Accessibility";
}

#pragma mark - Public API

// ---------------------------------------------------------------------------
// announce(message)
//
// Posts UIAccessibilityAnnouncementNotification so VoiceOver reads the message
// aloud. This is the iOS equivalent of Android's View.announceForAccessibility().
//
// JavaScript:
//   Ti.Accessibility.announce('Data refreshed');
// ---------------------------------------------------------------------------
- (void)announce:(NSString *)message
{
  if (message == nil || [message length] == 0) {
    NSLog(@"[WARN] Ti.Accessibility.announce called with empty message, ignoring.");
    return;
  }

  // Must be dispatched on the main thread — UIKit accessibility APIs are not thread-safe.
  TiThreadPerformOnMainThread(
      ^{
        UIAccessibilityPostNotification(UIAccessibilityAnnouncementNotification, message);
      },
      NO);
}

// ---------------------------------------------------------------------------
// focus(viewProxy)
//
// Posts UIAccessibilityLayoutChangedNotification with the resolved native view,
// causing VoiceOver to shift focus to that element and read it aloud.
//
// JavaScript:
//   Ti.Accessibility.focus(myRow);
// ---------------------------------------------------------------------------
- (void)focus:(id)viewProxy
{
  NSArray *currentArgs = [JSContext currentArguments];
  if (currentArgs != nil && [currentArgs count] > 0) {
    viewProxy = [currentArgs objectAtIndex:0];
  }

  if ([viewProxy isKindOfClass:[JSValue class]]) {
    viewProxy = [self JSValueToNative:(JSValue *)viewProxy];
  }

  if (![viewProxy isKindOfClass:[TiViewProxy class]]) {
    NSLog(@"[WARN] Ti.Accessibility.focus: argument must be a Ti.UI view proxy, got %@", [viewProxy class]);
    return;
  }

  TiViewProxy *proxy = (TiViewProxy *)viewProxy;

  // Resolving the view and posting the notification must happen on the main thread.
  TiThreadPerformOnMainThread(
      ^{
        if (![proxy viewAttached]) {
          NSLog(@"[WARN] Ti.Accessibility.focus: view is not yet realized, ignoring.");
          return;
        }

        // Retrieve the realized view
        TiUIView *uiView = [proxy view];

        // Retrieve the real accessibility element in case the view overrides it.
        id element = [uiView accessibilityElement];
        UIView *targetView = [element isKindOfClass:[UIView class]] ? element : uiView;

        // UIAccessibilityLayoutChangedNotification moves VoiceOver focus to the
        // supplied element on the next accessibility layout pass.
        // A slight delay ensures it is not swallowed if called during a window transition.
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.25 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
          UIAccessibilityPostNotification(UIAccessibilityLayoutChangedNotification, targetView);
        });
      },
      NO);
}

@end

#endif /* USE_TI_ACCESSIBILITY */

File: android/modules/accessibility/src/java/ti/modules/titanium/accessibility/AccessibilityModule.java
/**
 * Titanium SDK
 * Copyright TiDev, Inc. 04/07/2022-Present. All Rights Reserved.
 * Licensed under the terms of the Apache Public License
 * Please see the LICENSE included with this distribution for details.
 */
package ti.modules.titanium.accessibility;

import org.appcelerator.kroll.KrollModule;
import org.appcelerator.kroll.annotations.Kroll;
import org.appcelerator.kroll.common.Log;
import org.appcelerator.titanium.TiApplication;
import org.appcelerator.titanium.proxy.TiViewProxy;
import org.appcelerator.titanium.view.TiUIView;

import android.view.accessibility.AccessibilityEvent;
import android.view.accessibility.AccessibilityManager;
import android.view.accessibility.AccessibilityNodeInfo;
import androidx.core.view.accessibility.AccessibilityEventCompat;

@Kroll.module
public class AccessibilityModule extends KrollModule
{
	private static final String TAG = "AccessibilityModule";

	@Kroll.method
	public void announce(String message)
	{
		AccessibilityManager accessibilityManager = TiApplication.getInstance().getAccessibilityManager();
		if (!accessibilityManager.isEnabled()) {
			Log.w(TAG, "Accessibility announcement ignored. Accessibility services are not enabled on this device.");
			return;
		}
		if (message == null || message.isEmpty()) {
			Log.w(TAG, "Accessibility announcement ignored. No announcement text was provided.");
			return;
		}

		AccessibilityEvent event = AccessibilityEvent.obtain(AccessibilityEventCompat.TYPE_ANNOUNCEMENT);
		event.setEnabled(true);
		event.getText().clear();
		event.getText().add(message);
		accessibilityManager.sendAccessibilityEvent(event);
	}

	@Kroll.method
	public void focus(Object viewProxy)
	{
		if (!(viewProxy instanceof TiViewProxy)) {
			Log.w(TAG, "Ti.Accessibility.focus: argument must be a Ti.UI view proxy.");
			return;
		}
		
		TiViewProxy proxy = (TiViewProxy) viewProxy;
		TiUIView uiView = proxy.peekView();
		if (uiView == null) {
			Log.w(TAG, "Ti.Accessibility.focus: view is not yet realized, ignoring.");
			return;
		}

		final android.view.View nativeView = uiView.getNativeView();
		if (nativeView != null) {
			nativeView.post(new Runnable() {
				@Override
				public void run()
				{
					nativeView.performAccessibilityAction(AccessibilityNodeInfo.ACTION_ACCESSIBILITY_FOCUS, null);
				}
			});
		}
	}

	@Override
	public String getApiName()
	{
		return "Ti.Accessibility";
	}
}
